### ISSUE TITLE
captureException/captureError callback called before asynchronous function complete

### ISSUE THREAD

**Issue by colbygk** [getsentry/raven-node#130]
It appears that the captureException/captureError functions provide a facility to have a callback defined. The path of least surprise would suggest that that callback would be called only after the call to sending the message to sentry has been sent.  However, the call back appears to be called as soon as the message is placed on the async queue.

This is a problem when attempting to handle sending an error to sentry when catching an uncaughtException and sending a message just before exiting a node process.

**Comment by mattrobenolt** [comment]
This is intentional. The event system should be used to determine when the event was actually submitted.

Typically, you want to call `client.captureException` and get back the event id, then use the event id, before the message has actually been sent.

See: https://docs.getsentry.com/hosted/clients/node/usage/#events

**Comment by mattrobenolt** [comment]
Also, to be clear, there are two use cases here, which are a bit awkward. So I'm open to suggestions for something that makes more sense.

So there is one thing that can be done, but would require changing the API.

We can calculate the event id before doing anything async and return that from `capture*` methods. Then use the callback to determine when it's sent.

For example, that'd mean:

```javascript
var eventId = client.captureException(..., function(result, err) {
  // message was sent or something
});
```

But this will require a bit of restructuring and a major version bump.

**Comment by colbygk** [comment]
I think this runs counter to the actual intent behind callbacks in node (https://github.com/maxogden/art-of-node#callbacks) and that the handling of the callback really ought to be inside the function that received the callback in the first place, rather than pushing that onto the user of the library. It's good to hear that there's an openness to having it modified to react in a way that would be less surprising for node patterns.

**Comment by mattrobenolt** [comment]
Yeah, I understand.

So let me explain what's going on here:

The eventId is generated on the client. It's just a uuid4. This eventId is something that can be shown to the user.

Think about an HTTP framework, and you want to show this errorId on your error page. You don't want your user to wait to see the error page, if Sentry is being slow or sentry is down. So we absolutely attempt to do all actual sending to our servers async and this shouldn't affect your server more.

If for example, it were taking 5 seconds to send the message to us, we don't want to cause another 5 seconds wait to display your error page. This also means we're occupying now an open socket when things are obviously already having an issue. If Sentry were to add slowness to your page, this could cause a cascade of other issues. This isn't good.

So with that said, we want to get back an eventId ASAP, then opt into waiting for the actual response to be successful or not.

Historically, calculating the eventId required doing other async operations. (reading files from disk), so we couldnt' calculate the eventId prematurely, and we needed to use a callback just to get eventId.

Based on that decision, I chose to use an event listener for checking when events were actually sent since that's a less common use-case, and returned eventId in the callback.

Make sense?
